File: Interview Preparation Kit/Arrays/NewYearChaos.js
// https://www.hackerrank.com/challenges/new-year-chaos/problem
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

// Complete the minimumBribes function below.
function minimumBribes(q) {
    let swaps = 0;
    let index = 0;
    let lastSortedIndex = -1;
    let too_chaotic = false;
    let bribeMap = new Map();

    let stop = q.length - 1;
    while (index < stop) {
        let current = q[index];
        if (current > q[index + 1]) {
            if (bribeMap.has(current) && bribeMap.get(current) >= 2) {
                console.log("Too chaotic");
                too_chaotic = true;
                break;
            }
            if (!bribeMap.has(current)) {
                bribeMap.set(current, 0);
            }
            bribeMap.set(current, bribeMap.get(current) + 1);

            let temp = q[index];
            q[index] = q[index + 1];
            q[index + 1] = temp;

            swaps++;
            index = lastSortedIndex;
        }
        else if (current == index + 1) {
            lastSortedIndex++;
        }
        index++;
    }

    if (!too_chaotic) {
        console.log(swaps);
    }
}

function main() {
    const t = parseInt(readLine(), 10);

    for (let tItr = 0; tItr < t; tItr++) {
        const n = parseInt(readLine(), 10);

        const q = readLine().split(' ').map(qTemp => parseInt(qTemp, 10));

        minimumBribes(q);
    }
}

File: Interview Preparation Kit/Arrays/NewYearChaos.go
// https://www.hackerrank.com/challenges/new-year-chaos/problem
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

/*
 * Complete the 'minimumBribes' function below.
 *
 * The function accepts INTEGER_ARRAY q as parameter.
 */
func minimumBribes(q []int32) {
	swaps, index, lastSortedIndex := 0, int32(0), int32(-1)
	too_chaotic := false
	bribeMap := make(map[int32]int32)

	stop := int32(len(q) - 1)
	for index < stop {
		current := q[index]
		if current > q[index+1] {
			val, exists := bribeMap[current]
			if exists && val >= 2 {
				fmt.Println("Too chaotic")
				too_chaotic = true
				break
			}
			if !exists {
				bribeMap[current] = 0
			}
			bribeMap[current]++

			q[index], q[index+1] = q[index+1], q[index]

			swaps++
			index = lastSortedIndex
		} else if current == index+1 {
			lastSortedIndex++
		}
		index++
	}

	if !too_chaotic {
		fmt.Println(swaps)
	}
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	tTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
	checkError(err)
	t := int32(tTemp)

	for tItr := 0; tItr < int(t); tItr++ {
		nTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
		checkError(err)
		n := int32(nTemp)

		qTemp := strings.Split(strings.TrimSpace(readLine(reader)), " ")

		var q []int32

		for i := 0; i < int(n); i++ {
			qItemTemp, err := strconv.ParseInt(qTemp[i], 10, 64)
			checkError(err)
			qItem := int32(qItemTemp)
			q = append(q, qItem)
		}

		minimumBribes(q)
	}
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
